agent/lm_agent/parsing/olicense.py

```diff
@@ -0,0 +1,135 @@
+    Data we need:
+    - ``in_use``: quantity of licenses being used by each user
+
+    Obs: this line doesn't include the license name.
+    The license in use is the last one parsed before this line.
+    It also doesn't include the user name.
+    The user using the license is the next usage line parsed after this line.
+    This line is printed for each user using the license.
+    The total amount of licenses being used is the sum of all in use lines.
+    """
+
+    parsed_in_use = RX_IN_USE.match(line)
+    if parsed_in_use is None:
+        return None
+    in_use_data = parsed_in_use.groupdict()
+
+    return int(in_use_data["in_use"])
+
+
+def parse_usage_line(line: str) -> Optional[dict]:
+    """
+    Parse the usage line in the Olicense output.
+    Data we need:
+    -``user_name``: user who booked the license
+    -``lead_host``: host using the license
+
+    Obs: this line also doesn't include the license name.
+    The license in use is the last one parsed before this line.
+    It also doesn't include the quantity of licenses being used.
+    Since this line is printed for each license used by the user, the booked quantity is always 1.
+    """
+    parsed_usage = RX_USAGE.match(line)
+    if parsed_usage is None:
+        return None
+    usage_data = parsed_usage.groupdict()
+
+    return {
+        "user_name": usage_data["user"],
+        "lead_host": usage_data["lead_host"],
+        "booked": 1,
+    }
+
+
+def parse(server_output: str) -> dict:
+    """
+    Parse the OLicense output using regex to match the lines we need:
+    -``feature line``: info about each license
+    -``in use line``: info about licenses in use
+    -``usage line``: info about users using licenses
+
+    Since the in use and usage line don't have the name of the license in use,
+    we're saving each parsed license in a list. This way, we can find which
+    license is being used by checking the last parsed license in the list.
+
+    If a feature has more than one license associated with it, the total amount
+    of licenses is the sum of all licenses with the same name.
+    """
+    parsed_data: dict = {}
+    feature_list: list = []
+
+    for line in server_output.splitlines():
+        parsed_feature = parse_feature_line(line)
+        parsed_in_use = parse_in_use_line(line)
+        parsed_usage = parse_usage_line(line)
+
+        if parsed_feature:
+            feature = parsed_feature["feature"]
+            total = parsed_feature["total"]
+            if feature not in feature_list:
+                feature_list.append(feature)
+                parsed_data[feature] = {"used": 0, "total": total, "uses": []}
+            else:
+                parsed_data[feature]["total"] += total
+        elif parsed_in_use:
+            feature = feature_list[-1]
+            parsed_data[feature]["used"] += parsed_in_use
+        elif parsed_usage:
+            feature = feature_list[-1]
+            parsed_data[feature]["uses"].append(parsed_usage)
```

Comment:
For efficiency's sake, I think you shouldn't apply the other parsers if one already is successful. Maybe something like:
```suggestion
        parsed_feature = parse_feature_line(line)
        if parsed_feature:
            feature = parsed_feature["feature"]
            total = parsed_feature["total"]
            if feature not in feature_list:
                feature_list.append(feature)
                parsed_data[feature] = {"used": 0, "total": total, "uses": []}
            else:
                parsed_data[feature]["total"] += total
            continue
            
        parsed_in_use = parse_in_use_line(line)
        if parsed_in_use:
            feature = feature_list[-1]
            parsed_data[feature]["used"] += parsed_in_use
            continue
            
        parsed_usage = parse_usage_line(line)
        if parsed_usage:
            feature = feature_list[-1]
            parsed_data[feature]["uses"].append(parsed_usage)
```

Reply:
Thanks, I addressed it in the latest commit.